_posts/2023-09-06-load-balancer.md

```diff
@@ -0,0 +1,906 @@
+
+    sendAndExpect(backend)
+      .map { obtained =>
+        assertEquals(obtained, HttpServer.Status.Dead)
+      }
+      .unsafeRunSync()
+  }
+}
+```
+Cool!
+
+We need a few more services before we write the whole logic for load balancer or backend health checks, so, let's move on.
+
+Now we need a special service which will be responsible for applying Round Robin logic to `UrlsRef`-s, so to the `Backends` and `HealthChecks` which we defined in the `domain` part.
+```scala
+package com.ghurtchu.loadbalancer.services
+
+import cats.Id
+import cats.effect.IO
+import com.ghurtchu.loadbalancer.domain.Urls.Url
+import com.ghurtchu.loadbalancer.domain.{Backends, UrlsRef}
+
+trait RoundRobin[F[_]] {
+  def apply(ref: UrlsRef): IO[F[Url]]
+}
+
+object RoundRobin {
+
+  type BackendsRoundRobin     = RoundRobin[Option]
+  type HealthChecksRoundRobin = RoundRobin[Id]
+
+  def forBackends: BackendsRoundRobin = new BackendsRoundRobin {
+    override def apply(ref: UrlsRef): IO[Option[Url]] =
+      ref.urls
+        .getAndUpdate(_.next)
+        .map(_.currentOpt)
+
+  }
+
+  def forHealthChecks: HealthChecksRoundRobin = new HealthChecksRoundRobin {
+    override def apply(ref: UrlsRef): IO[Id[Url]] =
+      ref.urls
+        .getAndUpdate(_.next)
+        .map(_.currentUnsafe)
+  }
+
+  val testId: RoundRobin[Id] = new RoundRobin[Id] {
+    override def apply(ref: UrlsRef): IO[Id[Url]] =
+      IO.pure(Url("localhost:8081"))
+  }
+
+  val testOpt: RoundRobin[Option] = new RoundRobin[Option] {
+    override def apply(ref: UrlsRef): IO[Option[Url]] =
+      IO.pure(Some(Url("localhost:8081")))
+  }
+}
+
+```
+
+(I omitted the tests for `RoundRobin` because it's really huge, you can view whole the source code in the end)
+
+Nice! 
+
+It's time to create something that based on `Status` will update the `Backends` atomically - either remove or add new `Url` to it:
+
+```scala
+package com.ghurtchu.loadbalancer.services
+
+import cats.effect.IO
+import com.ghurtchu.loadbalancer.domain.Backends.Backends
+import com.ghurtchu.loadbalancer.domain.Urls.Url
+import com.ghurtchu.loadbalancer.domain.{Backends, Urls}
+import com.ghurtchu.loadbalancer.services.SendAndExpect.Status
+
+trait UpdateBackendsAndGet {
+  def apply(backends: Backends, url: Url, status: Status): IO[Urls]
+}
+
+object UpdateBackendsAndGet {
+
+  def impl: UpdateBackendsAndGet = new UpdateBackendsAndGet {
+    /**
+     * if status is Alive it adds url to backends atomically
+     * if status is Dead it removes url from backends atomically
+     */
+    override def apply(backends: Backends, url: Url, status: Status): IO[Urls] =
+      status match {
+        case Status.Alive =>
+          IO.println(s"$url is alive") *>
+            backends.urls
+              .updateAndGet(_.add(url))
+        case Status.Dead  =>
+          IO.println(s"$url is dead") *>
+            backends.urls
+              .updateAndGet(_.remove(url))
+      }
+  }
+}
+```
+
+Looks so simple, right? 
```

Comment:
Be _very_ careful when saying "this is so simple". I would avoid saying that here